packages/blade/src/components/ActionList/ActionList.tsx

```diff
@@ -1,119 +1,149 @@
-        selectInputRef.current?.focus();
-      }}
-      onMouseDown={() => {
-        setShouldIgnoreBlur(true);
-      }}
-      href={props.href}
-      style={{
-        border: activeIndex === props._index ? '2px solid red' : '',
-        width: '100%',
-      }}
-    >
-      <Text>{props.title}</Text>
-    </Box>
-  );
-};
 
-type ActionListProps = {
-  children: React.ReactNode[];
-};
-const ActionList = ({ children }: ActionListProps): JSX.Element => {
-  const { setOptions, actionListRef, selectionType, dropdownBaseId } = useDropdown();
-  const actionListOptions: {
-    title: string;
-    value: string;
-  }[] = [];
+      const clonedChild = React.cloneElement(child, {
+        // @ts-expect-error: TS doesn't understand the child's props
+        _index: currentIndex,
+      });
+      return clonedChild;
+    }
 
-  console.count('ActionList');
+    return child;
+  };
+
+  const childCompIdArray = React.Children.toArray(children).map((child) => getComponentId(child));
+  const lastActionListSectionIndex = childCompIdArray.lastIndexOf(componentIds.ActionListSection);
+
+  const isActionListItemPresentAfterSection = childCompIdArray
+    .slice(lastActionListSectionIndex)
+    .includes(componentIds.ActionListItem);
 
   // Looping through ActionListItems to add index to them and get an options array for moving focus between items
-  const childrenWithId = React.Children.map(children, (child) => {
+  const childrenWithId = React.Children.map(children, (child, index) => {
     if (React.isValidElement(child)) {
-      // @TODO: handle the scenario where ActionListItem is inside ActionListMenu
-      if (child.type === ActionListItem) {
-        actionListOptions.push({
-          title: child.props.title,
-          value: child.props.value,
-        });
-        const currentIndex = actionListOptions.length - 1;
-        const clonedChild = React.cloneElement(child, {
+      if (isValidAllowedChildren(child, componentIds.ActionListSection)) {
+        return React.cloneElement(child, {
           // @ts-expect-error: TS doesn't understand the child's props
-          _index: currentIndex,
+          children: React.Children.map(child.props.children, (childInSection) => {
+            if (isValidAllowedChildren(childInSection, componentIds.ActionListItem)) {
+              return getActionListItemChildWithId(childInSection);
+            }
+
+            return childInSection;
+          }),
+          hideDivider: index === lastActionListSectionIndex && !isActionListItemPresentAfterSection,
         });
-        return clonedChild;
       }
-    }
 
+      if (isValidAllowedChildren(child, componentIds.ActionListItem)) {
+        return getActionListItemChildWithId(child);
+      }
+    }
     return child;
   });
 
   React.useEffect(() => {
     setOptions(actionListOptions);
     // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [setOptions, children]);
+  }, [optionsRecalculateToggle]);
+
+  React.useEffect(() => {
```

Comment:
oh I am assuming these things to work after hydration only.

Even with isomorphic hook, it won't run on server. We use `useEffect` on server and `useLayoutEffect` on browser which might speed things up a bit. But in React docs they mostly suggested to use `useEffect` wherever possible. And the speed at which the options and selections load isn't that big of a deal right now. Its helpful when you want to load something quick on first fold of the page but in our case, we're only loading the options in the state.

Overall its not possible to run it on server because we're setting it into the state.

<img width="936" alt="image" src="https://user-images.githubusercontent.com/30949385/215381960-3ab6f7e8-23da-4561-9a1c-0594bc46dcc4.png">